keel-core/src/test/kotlin/com/netflix/spinnaker/keel/constraints/TimeWindowConstraintEvaluatorTests.kt

```diff
@@ -0,0 +1,168 @@
+package com.netflix.spinnaker.keel.constraints
+
+import com.netflix.spinnaker.keel.api.ArtifactType
+import com.netflix.spinnaker.keel.api.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.TimeWindow
+import com.netflix.spinnaker.keel.api.TimeWindowConstraint
+import com.netflix.spinnaker.keel.exceptions.InvalidConstraintException
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+import io.mockk.mockk
+import strikt.api.expectCatching
+import strikt.api.expectThat
+import strikt.assertions.failed
+import strikt.assertions.isA
+import strikt.assertions.isFalse
+import strikt.assertions.isTrue
+import java.time.Clock
+import java.time.Instant
+import java.time.ZoneId
+
+internal class TimeWindowConstraintEvaluatorTests : JUnit5Minutests {
+  data class Fixture(
+    val clock: Clock,
+    val constraint: TimeWindowConstraint
+  ) {
+    val artifact = DeliveryArtifact("fnord", ArtifactType.DEB)
+    val constrainedEnvironment = Environment(
+      name = "prod",
+      constraints = setOf(constraint)
+    )
+    private val previousEnvironment = Environment(
```

Comment:
Philosophical question for the group: I thought `time-window` would be a type of constraint/requirement that would gate deployment into an environment regardless of a promotion flow between environments (I remember discussing this with Michael for the MD blog post), but the function in the evaluator interface is called `canPromote`. Can this be used with a single environment?

Reply:
I think you should be able to control deployment into any environment with a `time-window` constraint. That applies if you have only 1 environment and you only want new builds deployed there on, let's say, weekdays.

I think that each constraint that is on an environment is evaluated, and they are && together. So if every constraint returns `true` to `canPromote()` then we would promote the artifact.

Reply:
Thanks, Emily. I was just wondering if for the tests for this particular constraint (unlike for `depends-on`, which obviously needs 2 environments), we should then have a single environment in the fixtures.

Reply:
I'll delete "staging" from the fixture if its causing confusion.